A couple of Makefile tweaks

Keno opened the pull request:
- If we're building the world with a custom C++ library, make sure
  PATCHELF picks up the correct flags both at build time and when
  we use it.
- Fix base/version_git.sh to also work when using O= out of tree
  builds (important for Compat, etc.)
- Don't link against the C++ ABI level in C-only dependencies (because it might not be built yet)

Keno committed:
A couple of Makefile tweaks

- If we're building the world with a custom C++ library, make sure
  PATCHELF picks up the correct flags both at build time and when
  we use it.
- Fix base/version_git.sh to also work when using O= out of tree
  builds (important for Compat, etc.)

tkelman commented:
Why does it matter which c++ std library patchelf gets built against?

Keno commented:
If you're building everything with sanitizer it'll fail otherwise. Of course, the alternative solution is to change to build rules to prevent patchelf from being built with sanitizers, but that seemed like an annoying special case.

tkelman commented:
Huh. Picky, but whatever. It's also used on arpack, I suspect you'll need to change `$(PATCHELF)` to `$(PATCHELF_BIN)` there too.

Keno committed:
Prevent infinite loop in relative path script

Keno committed:
Also use PATCHELF_BIN for ARPACK

Keno commented:
Thanks, not sure why I missed that one.

Keno committed:
Don't touch libuv's configure in it's build rule

It's the prerequisite for this rule, so touching it will invalidate the libuv for all other trees.

Keno committed:
Put quotes around common_part

yuyichao commented:
This also seems to fix the infinite loop in `contrib/relative_path.sh /usr/bin /etc` for me after https://github.com/JuliaLang/julia/pull/14097 .

vtjnash commented:
@tkelman 40386c9 shouldn't have been merged

tkelman commented:
whoops, sorry about that, we can revert
